File: src/main/java/com/linroid/knode/js/JSPromise.kt
package com.linroid.knode.js

/**
 * @author linroid
 * @since 2019/11/1
 */
class JSPromise : JSObject {

    private var resolverPtr: Long = 0

    @NativeConstructor
    private constructor(context: JSContext, reference: Long) : super(context, reference)

    constructor(context: JSContext) : super(context, 0) {
        nativeNew()
    }

    fun reject(error: String) {
        nativeReject(JSError(context, error))
    }

    fun resolve(value: Any?) {
        nativeResolve(from(context, value))
    }

    fun then(callback: (JSValue) -> Unit): JSPromise {
        val then = JSFunction(context, "then") { _, argv ->
            check(argv.isNotEmpty()) { "then() should receive a JSValue argument" }
            val result: JSValue = argv[0]
            callback.invoke(result)
            return@JSFunction null
        }
        nativeThen(then)
        return this
    }

    fun catch(callback: (JSError) -> Unit): JSPromise {
        val then = JSFunction(context, "catch") { _, argv ->
            val result: JSValue = argv[0]
            check(result is JSError) { "catch() should receive an JSError parameter not ${result.javaClass.simpleName}(${result.typeOf()})}: ${result.toJson()}" }
            callback.invoke(result)
            return@JSFunction null
        }
        nativeCatch(then)
        return this
    }

    private external fun nativeNew()
    private external fun nativeReject(error: JSError)
    private external fun nativeResolve(value: JSValue)

    private external fun nativeThen(callback: JSFunction)
    private external fun nativeCatch(callback: JSFunction)
}

File: knode/src/main/cpp/types/JSPromise.cpp
//
// Created by linroid on 2019/11/8.
//

#include "JSPromise.h"
#include "JSValue.h"
#include "JSError.h"

jclass JSPromise::jClazz;
jmethodID JSPromise::jConstructor;
jfieldID JSPromise::jResolver;

jint JSPromise::OnLoad(JNIEnv *env) {
  jclass clazz = env->FindClass("com/linroid/knode/js/JSPromise");
  if (!clazz) {
    return JNI_ERR;
  }
  JNINativeMethod methods[] = {
    {"nativeNew",     "()V",                                  (void *) JSPromise::New},
    {"nativeReject",  "(Lcom/linroid/knode/js/JSError;)V",    (void *) JSPromise::Reject},
    {"nativeResolve", "(Lcom/linroid/knode/js/JSValue;)V",    (void *) JSPromise::Resolve},
    {"nativeThen",    "(Lcom/linroid/knode/js/JSFunction;)V", (void *) JSPromise::Then},
    {"nativeCatch",   "(Lcom/linroid/knode/js/JSFunction;)V", (void *) JSPromise::Catch},
  };
  jClazz = (jclass) env->NewGlobalRef(clazz);
  jConstructor = env->GetMethodID(clazz, "<init>", "(Lcom/linroid/knode/js/JSContext;J)V");
  jResolver = env->GetFieldID(clazz, "resolverPtr", "J");
  env->RegisterNatives(clazz, methods, sizeof(methods) / sizeof(JNINativeMethod));
  return JNI_OK;
}

void JSPromise::New(JNIEnv *env, jobject jThis) {
  V8_SCOPE(env, jThis)
  auto context = runtime->context_.Get(isolate);
  auto resolver = v8::Promise::Resolver::New(context).ToLocalChecked();
  auto promise = resolver->GetPromise();
  auto resolverResult = new v8::Persistent<v8::Value>(isolate, resolver);
  auto promiseResult = new v8::Persistent<v8::Value>(isolate, promise);
  JSValue::SetReference(env, jThis, (jlong) promiseResult);
  env->SetLongField(jThis, jResolver, reinterpret_cast<jlong>(resolverResult));
}

void JSPromise::Reject(JNIEnv *env, jobject jThis, jobject jError) {
  auto value = JSValue::Unwrap(env, jError);
  jlong resolverPtr = env->GetLongField(jThis, jResolver);
  auto resolver = reinterpret_cast<v8::Persistent<v8::Promise::Resolver> *>(resolverPtr);
  SETUP(env, jThis, v8::Promise)
  v8::TryCatch tryCatch(runtime->isolate_);
  UNUSED(resolver->Get(isolate)->Reject(context, value->Get(isolate)));
  if (tryCatch.HasCaught()) {
    runtime->Throw(env, tryCatch.Exception());
    return;
  }
}

void JSPromise::Resolve(JNIEnv *env, jobject jThis, jobject jValue) {
  auto value = JSValue::Unwrap(env, jValue);
  jlong resolverPtr = env->GetLongField(jThis, jResolver);
  auto resolver = reinterpret_cast<v8::Persistent<v8::Promise::Resolver> *>(resolverPtr);
  SETUP(env, jThis, v8::Promise)
  v8::TryCatch tryCatch(runtime->isolate_);
  UNUSED(resolver->Get(isolate)->Resolve(context, value->Get(isolate)));
  if (tryCatch.HasCaught()) {
    runtime->Throw(env, tryCatch.Exception());
    return;
  }
  // runtime->isolate_->RunMicrotasks();
}

void JSPromise::Then(JNIEnv *env, jobject jThis, jobject jCallback) {
  auto callback = reinterpret_cast<v8::Persistent<v8::Function> *>(JSValue::GetReference(env, jCallback));
  SETUP(env, jThis, v8::Promise)
  v8::TryCatch tryCatch(runtime->isolate_);
  auto ret = that->Then(context, callback->Get(isolate));
  if (ret.IsEmpty()) {
    runtime->Throw(env, tryCatch.Exception());
    return;
  }
}

void JSPromise::Catch(JNIEnv *env, jobject jThis, jobject jCallback) {
  auto callback = reinterpret_cast<v8::Persistent<v8::Function> *>(JSValue::GetReference(env, jCallback));
  SETUP(env, jThis, v8::Promise)
  v8::TryCatch tryCatch(runtime->isolate_);
  auto ret = that->Catch(context, callback->Get(isolate));
  if (ret.IsEmpty()) {
    runtime->Throw(env, tryCatch.Exception());
    return;
  }
}
